push/ups_integration_with_openshift.md

```diff
@@ -0,0 +1,122 @@
+The developer creates a 'Mobile Client' in OpenShift as a logical construct to represent the App they are developing outside of OpenShift.
+In UPS, the developer creates a 'Push Application' and 1 or more 'Variants' to represent the App they are developing outside of OpenShift.
+There is duplication of logical constructs and potential confusion between what Mobile Client is and what a Push Application is.
+
+The developer also has to somehow get the Push configuration from UPS and place it into their App for consumption by the SDK. With the mobile-services.json approach being used by the Mobile CLI/UI and the SDKs for configuration, there's some things that can be done to simplify what the developer has to do here.
+
+*As a developer, I want to write custom logic for sending Push messages so that I can dynamically send messages based on varying factors and inputs*
+
+Push messages can be sent by calling the [sender API](https://aerogear.org/docs/specs/aerogear-unifiedpush-rest/index.html#397083935) with the correct key. There are features available in the OpenShift and the Catalog that can simplify the provisoining of a custom service, and obtaining secret credentials (like a key) from a provisioned service.
+
+## OAuth provider support 
+
+To address the first & second problem, UPS will be modified to support different auth providers.
+The 2 providers will be Keycloak (to maintain backwards compatibility for existing users) and a generic OAuth Provider.
+The OAuth flow will follow the same pattern & flow used by the [OpenShift OAuth Proxy](https://github.com/openshift/oauth-proxy).
+The UPS APB will need to be modified to include an OAuth ServiceAccount, similar to how [Jenkins uses a ServiceAccount](https://github.com/openshift/jenkins-openshift-login-plugin#browser-access).
+UPS will use this ServiceAccount as an OAuth Client to verify the User has the necessary permissions in OpenShift to use UPS.
+The permissions will be:
+* edit access on the namespace in OpenShift => Standard Developer Account in UPS
+* admin access on the namespace in OpenShift => Admin Account in UPS
+
+These changes will mean UPS will no longer have a dependency on Keycloak when running in OpenShift.
+Minimal changes will be required to the UPS APB.
+Also, the developer doesn't have to know about or be concerned with where or how the auth provider is setup and configured for UPS.
+
+## OpenShift Integration
+
+To address the third problem, a sidecar will be added to the UPS APB.
+This sidecar will provide background bi-directional syncing of resources and metadata between OpenShift and UPS.
+The sidecar will be written in Go and utilise a kubernetes client for watching & updating Kubernetes resources.
+It will use a SerivceAccount for auth when watching resources.
+The sidecar will also use a ServiceAccount to access the UPS Admin REST API for reading & updating UPS resources.
+
+### Syncing
```

Comment:
Users will need to pick one of the variants available in config. Sync logic could be complex and can lead to the bugs. I have limited knowledge on this, but could we use the same aproach as upstream: 

Let users to specify their variants in the code as this is something they need to anyway when more than one is present? 

As for IOS plist > than json for storing secrets.
Upstream is designed to use plist and moving to JSON may be a downgrade.

Changing format and api may also affect migration paths from previous versions and upstream to 5.0.

Reply:
TL;DR - just wanted to flag that there is the option to not do any sync, but I'm not sure if that something we want to do. I think it's essential to get json example bellow because from SDK point of view it's not obvious what this will look like.

Reply:
> TL;DR - just wanted to flag that there is the option to not do any sync, but I'm not sure if that something we want to do. I think it's essential to get json example bellow because from SDK point of view it's not obvious what this will look like.

@wtrocki @danielpassos Do you see value in providing UI in Openshift for creating the Variants instead? i.e. the Openshift Web Console has some of the UPS Admin UI parts added to it, and calls the UPS Management API.

Reply:
I think so. This is how it works today, and It works pretty well. I think to find a way to provide it on the mobile-config.json is not going to be easy.

Reply:
To extend Passos response I kinda feel that services running on openshift should have their own UI, however if we do not want to update or maintain current UI then this may be some option.  I see people just using current push UI the same way grafana is used. Maintaing two separate UI is not convenient. This will also allow us to reuse implementation tests and documentation from upstream.
We did not write our own keycloak dashboards etc.

[Unpopular opinion] Generally putting some end user facing UI into "Platform as service" solution seems to be strange and we should avoid that. Non service related elements like mobile apps etc. are ok.

Reply:
> [Unpopular opinion] Generally putting some end user facing UI into "Platform as service" solution seems to be strange and we should avoid that. What I will expect from OpenShift UI is to see some architecture components like services, mobile apps etc. Push configuration is too big stretch.

@wtrocki what is your view of the Jenkins integration in Openshift, where BuildConfigs, Builds & Build Pipelines are a UI in front of Jenkins Jobs, Job Runs and Jenkins Pipelines?

Reply:
> @matzew - Do we care to keep the API the same?

@wtrocki I'd think that - for iOS - `plist` (like we currently do) is more natrual for _true_ iOS devs

Reply:
> @wtrocki what is your view of the Jenkins integration in Openshift, where BuildConfigs, Builds & Build Pipelines are a UI in front of Jenkins Jobs, Job Runs and Jenkins Pipelines?

@david-martin @wtrocki I think there is generaly pros and cons w/ both - our own view could be more `aligned` w/ Openshift that way, just talking HTTP w/ the actual services. which is nice, but also people might be already very familiar w/ the existing UI (which is likely the case on Jenkins, Prometheus and Grafana) - push is a different beast here, since not that known/needed - the others are very generic and popular services

Reply:
we still can generate some mobile-services.plist file, no ? and make the iOS SDK aware of that ?